parameter registration / tagging system

benanne opened the pull request:
This is a PR to address #164. It should not be merged before #182 and #204 (and will require a rebase). I just wanted to get things going on this as well. Hopefully the rebase won't be too painful.

to do list:

- [x] extract `create_param()`, turn it into a helper function
- [x] add `Layer.add_param()`, `Layer.params`
- [x] replace `Layer.get_params()` (and remove `Layer.get_bias_params()`)
- [x] update all `Layer` subclasses to use `Layer.add_param()` instead of `Layer.create_param()` and supply the correct tags, also remove any `Layer.get_params()` and `Layer.get_bias_params()` implementations
- [x] update helper functions that get parameters to support tags
- [x] update all the tests: move `create_param` tests, add `Layer.add_param` tests, add `Layer.get_params` tests, add additional tests for using helper functions with tags
- [x] clean up (remove the commented code)
- [x] rebase after #182 is merged

after rebasing:

- [x] add deprecation placeholders
- [x] add (correctly formatted) docstrings for `Layer.add_param`, `Layer.get_params`

after rebasing again (to include all docs changes so far):

- [x] update docstrings for helper functions to mention the tags
- [x] update documentation and examples (if necessary)

benanne commented:
I think I'm more or less done with this, I just need to add / update a bunch of documentation. But since there is some overlap with #204 there, I will wait to spend time on that until #182 and #204 have been merged and I'm able to rebase this.

benanne commented:
Intersting Python 2/3 difference here: in Python 3, the `keys()` method of an `OrderedDict` returns a `KeysView`, not a list. This is now what `get_params()` returns, and this causes a bunch of problems.

The easiest fix is just to wrap it in `list(...)` to make sure `get_params()` returns a list in both Python 2 and Python 3. I'm okay with that solution, but before doing that I figured I'd ask if there are any caveats here, or if there is maybe a better solution. I have little to no experience with Python 3 so I wouldn't know.

dnouri commented:
I also don't know a better way to do this other than using `list(d.keys())` or `list(d)`.

benanne commented:
Adding that fix makes it pass Travis, which is nice! There is no point in reviewing this yet though (if anyone was planning on doing that), better to wait until after the imminent rebase.

benanne commented:
currently rebasing this... it's proving to be a great git exercise!

benanne commented:
Done! I hope I did everything right. Tests seem to pass now so that's good.

f0k commented on lasagne/layers/base.py:
```diff
@@ -38,46 +36,33 @@ def __init__(self, incoming, name=None):
-        return []
+        result = list(self.params.keys())
 
-    def get_bias_params(self):
```
The base class should probably have a deprecated placeholder for now:
```python
def get_bias_params(self):
    warnings.warn(...)
    return self.get_params(regularizable=False)
```

f0k commented on lasagne/layers/helper.py:
```diff
@@ -11,8 +12,6 @@
     "get_output_shape",
     "get_all_params",
-    "get_all_bias_params",
-    "get_all_non_bias_params",
```
Again, we might want to have deprecated redirections for those.

f0k commented on lasagne/tests/layers/test_base.py:
```diff
@@ -27,41 +32,69 @@ def test_layer_from_shape(self, layer_from_shape):
     def test_named_layer(self):
         from lasagne.layers.base import Layer
         l = Layer(Mock(), name="foo")
         assert l.name == "foo"
```
Shouldn't this use the new `named_layer` fixture?

f0k commented on lasagne/tests/layers/test_helper.py:
```diff
@@ -4,7 +4,7 @@
 
 
-class TestGetAllLayers:
+class TestGetAllLayers(object):
```
Hmm, most of the tests have old-style classes... we should probably be consistent there. Is there any reason for using one or the other with pytest?

benanne replied:
I do it by default nowadays. I guess old-style classes are frowned upon? I don't know :) Don't really care either way, as long as all classes in the library itself are new-style.

f0k commented on lasagne/utils.py:
```diff
@@ -204,3 +204,76 @@ def compute_norms(array, norm_axes=None):
+            * a numpy array with the initial parameter values
+            * a Theano shared variable representing the parameters
+            * a function or callable that takes the desired shape of
+              the parameter array as its single argument.
```
"and returns a numpy array"

f0k commented on lasagne/utils.py:
```diff
@@ -204,3 +204,76 @@ def compute_norms(array, norm_axes=None):
+        if not isinstance(arr, np.ndarray):
+            raise RuntimeError("cannot initialize parameters: the "
+                               "provided callable did not return a numpy "
+                               "array")
```
This is independent of this PR, but as I saw it now, this check should probably go something like this instead:
```python
arr = spec(shape)
try:
    arr = floatX(arr)
except Exception:
    ... did not return an array-like ...
if arr.shape != shape:
    ... did not return correct shape ...
return theano.shared(arr, name=name)
```
This should also handle scalars correctly.

benanne replied:
If I include this change in this PR then I should rebase it to include the new floatX implementation, I guess.

f0k commented:
Looks good so far! Cool that https://github.com/Lasagne/Lasagne/issues/164#issuecomment-84996815 seems to have been free of errors :)

benanne commented:
I've added deprecation placeholders and addressed your other comments, except the change to `create_param` to make it support scalar values. That depends on your recent change of the `floatX()` implementation, so I would have to rebase this PR first to add that in here. Maybe it's better kept as a separate thing anyway.

f0k commented:
Looks good! If you need help with the remaining TODOs, let me know.

benanne commented:
I should be able to finish it today. I'll go ahead and rebase it again as well so we can include the modifications to `create_param`.

benanne commented:
Looks like I need to wait for #204 to do the last two things, else I'll just be doing double work (or the rebase/merge will be painful).

benanne commented:
Rebased once more so I can update the docs later :)

f0k commented on lasagne/layers/base.py:
```diff
@@ -39,49 +37,56 @@ def __init__(self, incoming, name=None):
+        By default, all parameters that were registered with
+        ``Layer.add_param()`` will be returned, in the order they were
+        registered. The list can optionally be filtered by specifying tags
+        as keyword arguments.
```
I'm wondering if this should be targetted more towards users of layers, who don't know about `add_param`. An alternative might be:
```
By default, all parameters that participate in the forward pass will be returned (in the
order they were registered in the Layer's constructor via :meth:`add_param()`). The list
can optionally be filtered by specifying tags as keyword arguments. For example,
``trainable=True`` will only return trainable parameters, and ``regularizable=True``
will only return parameters that can be regularized (e.g, by L2 decay).
```
Feel free to mix and match if you agree.

benanne replied:
Sure, looks good.

benanne commented:
I've updated the remaining docs, and in one go I fleshed out the custom layers documentation a bit as well, since I had to update it anyway.

f0k commented on docs/modules/layers.rst:
```diff
@@ -7,6 +7,8 @@
 ----------------
 
+.. autofunction:: get_output
+.. autofunction:: get_output_shape
```
Ah, good catch.

f0k commented on docs/user/custom_layers.rst:
```diff
@@ -58,21 +60,100 @@ input as follows:
+
+    class DotLayer(lasagne.layers.Layer):
+        def __init__(self, num_units, W=lasagne.init.Normal(0.01), **kwargs):
+            super(DotLayer, self).__init__(**kwargs)
```
You're missing the `incoming` argument!

benanne replied:
whoops :)

f0k commented on docs/user/custom_layers.rst:
```diff
@@ -58,21 +60,100 @@ input as follows:
+
+Some layers can have multiple behaviors. For example, a layer implementing
+dropout should be able to be switched on or off. During training, we want it
+to apply dropout noise to its input, but during evaluation we don't want it to
```
*and upscale the surviving values

(The scaling should be mentioned somewhere to not confuse people familiar with the dropout paper.)

benanne replied:
Ah yes, I guess so. It's in the code anyway so we might as well explain it :)

f0k commented on docs/user/layers.rst:
```diff
@@ -143,10 +143,60 @@ Propagating data through layers
+
+However, this is **not recommended**! Some network layers may have
+non-deterministic output, such as dropout layers. If you compute the network
+outputs with separate calls to :func:`lasagne.layers.get_output()`, they will
```
*If you create the output expressions

(I think it's important not to mix up the actual, nonsymbolic computation with the creation of the symbolic expressions.)

f0k commented on lasagne/layers/base.py:
```diff
@@ -39,49 +37,58 @@ def __init__(self, incoming, name=None):
+        via :meth:`add_param()`). The list can optionally be filtered by
+        specifying tags as keyword arguments. For example, ``trainable=True``
+        will only return trainable parameters, and ``regularizable=True``
+        will only return parameters that can be regularized (e.g, by L2 decay).
```
*e.g.,

f0k commented on lasagne/layers/helper.py:
```diff
@@ -312,9 +311,9 @@ def get_output_shape(layer_or_layers, input_shapes=None):
+    This function gathers all parameters of all layers below one or
     more given :class:`Layer` instances, including the layer(s) itself. Its
     main use is to collect all parameters of a network just given the output
     layer(s).
```
I'd suggest to add the additional explanation on filtering from `Layer.get_params()` here as well, because `get_all_params()` is probably what people will learn about first.

f0k commented on lasagne/utils.py:
```diff
@@ -205,3 +205,82 @@ def compute_norms(array, norm_axes=None):
+    spec : numpy array, Theano shared variable, or callable
+        One of three things:
+            * a numpy array with the initial parameter values
+            * a Theano shared variable representing the parameters
```
I think for the markup to work there must be an new line before the list, and the list should not be intended (i.e., it should start on the same level as "One of three things").

f0k commented:
Almost there! As always, everything reads well.

benanne commented:
Thanks! It's surprising how easy it is to write docs when I'm doing it to avoid writing my dissertation instead ;)

f0k commented:
"lasagne/utils.py:220:1: W293 blank line contains whitespace" -- almost as bloody a beginner's mistake as mine earlier today ;)

benanne commented:
I've had that happen at least five times today. Sublime Text indents by default if you press enter. There must be a plugin that removes whitespace on blank lines on save or something :)

f0k commented:
It's the same with most editors, but the better ones have pep8 plugins that highlight all errors. What saved me with simpler editors was `git diff`, it highlights trailing white space in red (if you enabled colored diffs).

benanne commented:
I actually have a pep8/flake8 plugin for it but shortly after I installed it I turned off a few rules that were really doing my head in. 'Blank line contains whitespace' must have been one of them :p

benanne commented:
I'll leave it open till tomorrow afternoon in case there are any further comments.

skaae commented:
I think you forgot to add `get_all_bias_params` and `get_all_non_bias_params` to `__all__` in `helper.py`

benanne commented:
Looks like I did, good catch! I'll quickly fix it in master.

skaae commented:
btw I have question about add params.
In the old recurrent layer get params was

```
    def get_params(self):
        params = (helper.get_all_params(self.input_to_hidden) +
                  helper.get_all_params(self.hidden_to_hidden))
```

which i can nearly reproduce if i use
`get_all_bias_params` and `get_all_non_bias_params` and then add the parameters with the appropriate tags, but then i'll lose any extra tags that the user might have added. Can we add a helper function that returns the parameters and the tags for this situation?

benanne commented:
What about something like this?
```
    def get_params(self, **tags):
        params = (helper.get_all_params(self.input_to_hidden, **tags) +
                  helper.get_all_params(self.hidden_to_hidden, **tags))

        if self.learn_init:
            return params + self.get_init_params(**tags)
        else:
            return params
```

You should not use `get_all_bias_params` or `get_all_non_bias_params`, they are deprecation placeholders and will be removed.

skaae commented:
great, thanks

f0k commented:
I'd go a different route and modify `self.params` instead of overriding `get_params`:
```python
def __init__(...):
    ...
    self.params.update(self.input_to_hidden.params)
    self.params.update(self.hidden_to_hidden.params)
    initsomething = self.add_param(initsomething, initshape, trainable=learn_init)
```
That's way easier than checking the tags again in `self.get_init_params`, and my suggested implementation for the tagging system explicitly aimed at making `self.params` easily modifiable.

skaae commented:
I dont entirely get your second example. I have a related problem. 

In the LSTM layer i would like to optionally allow the user to provide an output network. The outptut network takes the current hidden state in the LSTM and calculates some output that is then used as input in the next timestep. This is typically used in language models.

The LSTM `get_params` needs to return all parameters that are registered for the LSTM and return the parameters of the `output_network` if it exists. 

I would like to be able to call `get_all_params` on self er something like that :)

I could use Sanders solution but then I would need to explicitly return all LSTM parameters.
At the moment I copied `get_params` from `base.py` and modified it to return output_network parameters, but that is pretty ugly.

f0k commented:
> I dont entirely get your second example.

Which part? It uses the base class implementation to return the `initsomething` parameter depending on the tags (I don't know what you called that parameter, that's why I referred to it as `initsomething`), and the `get_all_params` helper function to return the parameters of the embedded networks depending on the tags. Sander's suggestion was meant to do the same, but it failed to check the tags.

> I would like to be able to call  get_all_params  on self er something like that :)

That's what `super(name_of_the_current_class, self).get_params(**tags)` does.
